### PR TITLE
kiss: use blake3 for checksums

### PR DESCRIPTION
**Pull request by dylanaraps**
NOTE: I have tested this change locally and it works fine. Please try it
and let me know if you encounter any issues.

With this commit, the package manager will use blake3 (b3sum) for all
checksum verification. The rationale for this change is one of
performance with blake3 beating out sha256 in every case.

This requires that the repository package for the package manager now
depend on b3sum (https://git.sr.ht/~mcf/b3sum (my option) or
https://github.com/BLAKE3-team/BLAKE3). I recommend this be statically
linked.

This change is intentionally unintrusive as it is transitional. Existing
checksums will continue to be verifiable and no changes are needed for
things to continue to work as usual.

Any newly generated checksums will use blake3. As packages have their
checksums updated, sha256 will slowly but seamlessly disappear.

There is a little nuance to the way this transition process works.

1. Existing checksum files are called 'checksums'.
2. New checksum files are called 'checksum'.
3. Existing etcsum files are called 'etcsums'.
4. New etcsum files are called 'etcsum'.

This is the ~~best~~ only way I found to differentiate. Happy to hear other 
ideas. I also personally prefer the new naming to the old one (superfluous plural).

Once sha256 checksums have vanished, the package manager's code can be
further simplified and we can drop the multiple sha256 implementation nonsense.

Benchmarks:

-> du -sh linux-5.12.6.tar.zst
170.6M  linux-5.12.6.tar.zst
-> time sha256sum linux-5.12.6.tar.zst >/dev/null
    0m01.11s real     0m00.98s user     0m00.13s system
-> time b3sum linux-5.12.6.tar.zst >/dev/null
    0m00.14s real     0m00.11s user     0m00.02s system

-> du -sh wlroots-0.13.0.tar.zst
520.0K  wlroots-0.13.0.tar.zst
-> time sha256sum wlroots-0.13.0.tar.zst >/dev/null
    0m00.00s real     0m00.00s user     0m00.00s system
-> time b3sum wlroots-0.13.0.tar.zst >/dev/null
    0m00.00s real     0m00.00s user     0m00.00s system

-> du -sh llvm-12.0.0.src.tar.zst
61.8M   llvm-12.0.0.src.tar.zst
-> time sha256sum llvm-12.0.0.src.tar.zst >/dev/null
    0m00.43s real     0m00.38s user     0m00.05s system
-> time b3sum llvm-12.0.0.src.tar.zst >/dev/null
    0m00.05s real     0m00.05s user     0m00.00s system

### CODE REVIEW

File: kiss [hunk 1]
```diff
@@ -791,25 +802,23 @@ pkg_checksums() {
 pkg_verify() {
     # Verify all package checksums. This is achieved by generating a new set
     # of checksums and then comparing those with the old set.
-    verify_cmd="
-    NR==FNR{a[\$1];next}/^git  .*|^SKIP$/{next}!((\$1)in a){exit 1}
-    "
```
Comment (dylanaraps) [inline kiss:796]: Small unrelated change (`^git` should no longer appear in any checksums files). This was an old transitional change.

**Comment by konimex** [comment]
Reconsidering what I said in #40, say, is it possible to have multiple files of checksums (with different algorithms of course)? I'm talking about `checksums-sha256`, `checksums-blake3`, `checksums-blake2`, `checksums-sha512` (same thing with etcsums).

And the most important thing, is my proposed approach pointless?

**Comment by dylanaraps** [comment]
We didn't do that originally for the purposes of *enforcing* a single checksum algorithm. ie, no point in adding a suffix if all checksums are sha256.

**Comment by dylanaraps** [comment]
We could also shorten to `sum` if we were so inclined.

File: kiss (hunk 1 above)
Comment (dilyn-corner) [reply kiss:796]: The problem here is that many packages still haven't updated since the last checksums change.

Comment (dylanaraps) [reply kiss:796]: This should be fixable with a little script run over the repositories. (ie, delete all lines in checksums files that == `^git`). I didn't realize we still had packages using the old format.

Comment (dylanaraps) [reply kiss:796]: It's been almost a year since the code I removed was added (for reference): https://github.com/kisslinux/kiss/commit/d6917d5fe1558d0690a7dd001996d34540ed79d9

**Comment by cemkeylan** [comment]
Personally, I'm in favour of naming it `blake3sum` (or any name scheme that implies the algorithm/variant). As you have said, the `checksums` name was chosen to enforce a single algorithm, and if we are to move away from the current algorithm we must think about the possibility of moving away from that one in the future as well. Even if this is an extremely rare situation, I think that this change might cause quite a lot of confusion.

**Comment by cemkeylan** [comment]
Another idea would be adding headers on top of the `checksums` files for the hashing algorithm used to generate the `checksums` file. If the file does not have that header, we know that it's sha256. The upside is that we don't need to rename the file, so this change would be much more seamless, and would cause much less confusion. Here is an example:

```
BLAKE3
3ec2d0fb04bb7c69ce39e58e040fa5e579818ec8738ff6d94b2bda0399f931c7  Python-3.9.2.tar.xz
40700e9f984ec99c43a314884ef5863b0c2ef48b714f5a9a9057555dfc609c45  patches/something.patch
```

**Comment by dylanaraps** [comment]
That changes the file format and implies some kind of permanent structure for hash types. Again, this is contrary to how things are designed to work (single hashing algorithm). This PR is merely about changing the algorithm in a transitional way. ie, both are
*temporarily* supported until the previous algorithm is gone.

Nothing prevents us from doing a "harder" transition whereby we mark a date, do some preparation (install b3sum early, etc) and then mass-update the checksums files. Just an idea, I much prefer the current PR.

**Comment by dilyn-corner** [comment]
`checksums` should just contain sum(s), potentially the name of the file, though it's not necessary once you've read... what, a single line in the package manager documentation? The type of sum it is is irrelevant, I don't see a reason for it to not be opaque to the user what made it (opaque in the sense that the file won't tell them; it would of course be documented *somewhere*). 

The only reason to change how checksums are generated is 1) security/collisions/verification, and 2) performance. Verifying downloads is important but to my knowledge b3 and sha256 are effectively the same for our purposes. Because the performance increase is insanely high, that's a fine reason to change (though as has been pointed out, it offers little benefit in the "grand scheme" for users with slow download speeds), and thus there's basically zero downside.

I like `sum` for the same reason @dylanaraps like `checksum`, though it is less obvious what that file may be. 

My only "concern" is where `b3sum` will live. As mentioned, `busybox` doesn't include a b3 utility. We could shoehorn it in, perhaps. Or just make it a standalone in `repo/core`, OR bundle it with the package manager itself. I like the first two more than the last one (I'd rather see `kiss` thrown into a `busybox`), but having a standalone b3 makes the most sense to me. (This concern extends to #40, and I have the same conclusion).

**Comment by dylanaraps** [comment]
@dilyn-corner The package manager already depends on `curl` via `git` which is why curl is currently omitted. It should be explicitly listed regardless as `git` may be statically linked (and there's therefore no curl).

**Comment by dilyn-corner** [comment]
Yes, that's what I mean; `git` has been static for a while, and I imagine a few users are witnessing `curl` hit their `kiss-orphans` list at one point or another. Removing it is a little unforgiving :v

**Comment by cemkeylan** [comment]
I am not against for using blake3 by all means, but I kind of dislike your changes to implement it. My main problem is with the name `checksum`.

> I also personally prefer the new naming to the old one (superfluous plural).

I don't think that it's a superfluous plural, `checksums` can contain the checksum of multiple files. Same way the file containing the sources are called `sources` instead of `source`, even when 80% of the time it contains a single source.

> That changes the file format and implies some kind of permanent structure for hash types. Again, this is contrary to how things are designed to work (single hashing algorithm).

Isn't this patch already changing how things were designed to work? The idea of using the algroithm name on the file makes it clear what the change is, while the idea of using headers and sticking with `checksums` doesn't cause any confusion as the package structure is not changed and it's also not that hard to parse the header.

I'm not saying my ideas here are great, but I personally think they are better than renaming to `checksum`. What if you find a much better algorithm two years from now and decide to change the algorithm, will you rename the file to `checksums` again? Renaming the file to contain the algorithm name doesn't mean you can't be enforcing a single algorithm in the package manager. I think it's better to enforce a single algorithm as well, but I also think that the changes here should be as clear as possible.

**Comment by dylanaraps** [comment]
Please do ignore this PR if it's undesirable. If it *is* desirable, there's no rush on merging it. Take as much time as needed. Same goes for the pax PR. I just found the changes interesting/beneficial so I sent them this way.  :+1:

**Comment by cemkeylan** [comment]
I'm sorry, my intention wasn't to discourage you, I agree that both blake3 and pax is beneficial, I only thought that the naming of the `checksums` file could be altered. Otherwise I think that they are good ideas overall.

**Comment by dylanaraps** [comment]
I am not discouraged in the slightest. :)

**Comment by konimex** [comment]
Since this PR will affect the package *format*, I'm inclined to agree with Cem's first suggestion (name it `blake3sums` to account possible algo change in the future rather than transitioning to `checksum` or even `sum` because while it is concise, it's inaccurate for packages with multiple sources, but at this point I'm probably just being a pedant). Yes, the `kiss` package manager can (and perhaps *should*) enforce a single checksum algorithm, but the package *format* shouldn't be affected by the details.

**Comment by dilyn-corner** [comment]
```
https://github.com/BLAKE3-team/BLAKE3
    0m00.46s real     0m01.98s user     0m00.30s system
https://git.sr.ht/~mcf/b3sum
    0m02.38s real     0m02.22s user     0m00.16s system
sha256sum
    0m04.33s real     0m04.18s user     0m00.17s system
```

oofa doofa

**Comment by konimex** [comment]
```
https://github.com/BLAKE3-team/BLAKE3
     0m00.46s real     0m01.98s user     0m00.30s system 
```

Is this one the Rust version or the C version? (Since there are two source codes there)

**Comment by dilyn-corner** [comment]
That would be the rust version, specifically the binary they provide.

**Comment by dilyn-corner** [comment]
Ultimately this was determined to be a problem when using `lld`. Depending on what happens in the future, the below patch to mcf's `b3sum` will make it zoom just as quick as the rust impl.

```
diff --git a/blake3_cpuid.S b/blake3_cpuid.S
index 491be8f..f6b72ba 100644
--- a/blake3_cpuid.S
+++ b/blake3_cpuid.S
@@ -20,7 +20,7 @@ blake3_xgetbv:
 	xgetbv
 	ret
 
-.section .ctors,"aw"
+.section .init_array,"aw"
 .align	8
 .quad	blake3_detect_cpu_features
 #endif
```
